### PR TITLE
EVG-1758 fetch a distro teardown script by id

### PR DESCRIPTION
**Pull request by StuartButler**

### CODE REVIEW

File: rest/model/distro.go [hunk 1]
```diff
@@ -133,3 +133,37 @@ func (e *APIExpansion) ToService() (interface{}, error) {
 
 	return interface{}(d), nil
 }
+
+// APIDistroSetup is derived from a service layer set-up script (string)
```
Comment (tychoish) [inline rest/model/distro.go]: I think you can just use the API distro struct as it has a Setup field with the right type and tag...

File: rest/route/distro.go [hunk 2]
```diff
@@ -17,33 +17,229 @@ import (
...
+
+// Run returns the given distro's setup script.
+func (h *distroIDGetSetupHandler) Run(ctx context.Context) gimlet.Responder {
+	distro, err := h.sc.FindDistroById(h.distroID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for find() by distro id '%s'", h.distroID))
+	}
+
+	script := &model.APIDistroSetup{}
+	if err = script.BuildFromService(*distro); err != nil {
```
Comment (tychoish) [inline rest/route/distro.go]: I think you you want to reverse the pointers here... and make your buildfromservice methods take pointers (generally)

File: rest/route/distro.go [hunk 3]
```diff
@@ -17,33 +17,229 @@ import (
...
+	}
+}
+
+// Parse fetches the distroId and JSON payload from the http request.
+func (h *distroIDChangeSetupHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.distroID = gimlet.GetVars(r)["distro_id"]
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	if err := util.ReadJSONInto(body, h); err != nil {
```
Comment (tychoish) [inline rest/route/distro.go]: should we have the read the body into the distro struct which we have as a field in the handler type?

File: rest/route/distro.go [hunk 4]
```diff
@@ -17,33 +17,229 @@ import (
...
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for find() by distro id '%s'", h.distroID))
+	}
+
+	d.Setup = h.Setup
+	if err = h.sc.UpdateDistro(d); err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for update() by distro id '%s'", h.distroID))
+	}
+
+	apiDistro := &model.APIDistro{}
+	if err = apiDistro.BuildFromService(*d); err != nil {
```
Comment (tychoish) [inline rest/route/distro.go]: change pointers.

File: rest/route/distro.go [hunk 5]
```diff
@@ -17,33 +17,229 @@ import (
...
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "API error converting from distro.Distro to model.APIDistro"))
```
Comment (tychoish) [inline rest/route/distro.go]: this is probably a 400 not a 500

File: rest/route/distro.go [hunk 6]
```diff
@@ -17,33 +17,229 @@ import (
...
+
+// Run returns the given distro's teardown script.
+func (h *distroIDGetTeardownHandler) Run(ctx context.Context) gimlet.Responder {
+	distro, err := h.sc.FindDistroById(h.distroID)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "Database error for find() by distro id '%s'", h.distroID))
+	}
+
+	script := &model.APIDistroTeardown{}
+	if err = script.BuildFromService(*distro); err != nil {
```
Comment (tychoish) [inline rest/route/distro.go]: change pointers

File: rest/route/distro.go (hunk 3 above)
Comment (StuartButler) [reply rest/route/distro.go]: As discussed

File: rest/route/distro.go (hunk 5 above)
Comment (StuartButler) [reply rest/route/distro.go]: Leaving as a 500 per our discussion

File: rest/model/distro.go [hunk 7]
```diff
@@ -31,47 +31,52 @@ type APIDistro struct {
...
-			if err != nil {
-				return err
-			}
-			apiDistro.ImageID = ToAPIString(ec2Settings.AMI)
-		}
-		if v.ProviderSettings != nil {
-			apiDistro.ProviderSettings = *v.ProviderSettings
+		d = h.(distro.Distro)
+	case *distro.Distro:
+		d = *h.(*distro.Distro)
```
Comment (tychoish) [inline rest/model/distro.go]: I think this can be d = v

**Pull request merged**